sources/host-ctr/cmd/host-ctr/pid_namespaces.go

```
@@ -0,0 +1,184 @@
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// Get file descriptor that refers to a parent namespace, ioctl() request '_IO(NSIO, 0x2)'
+	nsGetParentNs        = 0xb702
+	lockfile      string = "/run/lock/host-ctr-pid-ns.lock"
+	manifestPath  string = "/etc/host-containers/info/pid-namespaces.json"
+)
+
+func getTaskPidNs(task *containerd.Task) (int64, error) {
+	pid := (*task).Pid()
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	_, err = unix.IoctlGetInt(pidNsFd, nsGetParentNs)
+	if err == syscall.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb syscall.Stat_t
+	if err := syscall.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+func addPidNsToManifest(pidNs int64) error {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	locked, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	if !locked {
+		return errors.New("failed to acquire lock for writing to PID NS manifest")
+	}
+	defer func() {
+		_ = fileLock.Unlock()
```

Should you handle the error here, and print a message? If not, then why not just this `defer fileLock.Unlock()`?

---

+1 - we should surface the error that this may raise from the nested `syscall.Flock` this attempts to make

---

I originally had `defer fileLock.Unlock()` and the linter complained about throwing away the error. I'll try returning it.

---

Looked alittle at this, and it seems that returning errors out of `defer` isn't super simple or elegant.

Maybe the easiest thing for us to do is just to call `fileLock.Unlock()` ourselves and capture the error.